includes/class-wc-payments-payment-request-button-handler.php

```diff
@@ -150,7 +150,10 @@ public function get_total_label() {
 	 * @return void
 	 */
 	public function set_session() {
-		if ( ! $this->is_product() || ( isset( WC()->session ) && WC()->session->has_session() ) ) {
+		if (
+			! ( $this->is_product() && $this->is_available_at( 'product' ) )
```

Comment:
@ricardo I've noticed there is another condition for product page in `should_show_payment_request_button` method [here](https://github.com/Automattic/woocommerce-payments/blob/d48df11bcc2c50ce931a699371390eb7ef7a2cac/includes/class-wc-payments-payment-request-button-handler.php#L431-L435), should we take that into account as well? So this line would look like this:

```
! ( $this->is_product() && $this->should_show_payment_request_button() )
```

IMO that would keep all the logic in the same place.

Reply:
Another thing worth doing here is to add a comment explaining why there is such a special treatment for product page. Something like: `Don't set sessions cookies on product pages to allow caching when payment request buttons are disabled. But keep cookies if there is already active WC session in place.`

Reply:
Good call. Makes sense to use `should_show_payment_request_button`.

Made the change in [8ff6ba9](https://github.com/Automattic/woocommerce-payments/pull/4085/commits/8ff6ba97e2cc006684775da76599718bab2c7d3a).